### PR TITLE
Generic associated data

### LINKED ISSUE #99: Make additional_data into a generic type
This way, we don't need to pay the price for deserialization when accessing it on the same process.

### PR DESCRIPTION
**Pull request by tautropfli**
Resolves #99 

Todo:
- [x] Revert ead0b2a and use released mockiato version

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -4,7 +4,7 @@
 
 #![feature(specialization)]
 #![feature(non_exhaustive)]
-#![feature(box_syntax)]
+#![feature(box_syntax, trait_alias)]
```
Comment (janhohenheim) [inline src/lib.rs]: Put the `trait_alias` feature on an own line

File: Cargo.toml [hunk 2]
```diff
@@ -22,11 +22,10 @@ keywords = [ "simulation", "2d", "physics" ]
 myelin-geometry = "2.4"
 nalgebra = "0.18"
 ncollide2d = "0.19"
-serde = "1.0"
-serde_derive = "1.0"
+serde = { version = "1.0", features = ["derive"] }
 alga = "0.9"
 nameof = "1.0"
-mockiato = { version = "0.4", optional = true }
+mockiato = { git = "https://github.com/myelin-ai/mockiato", optional = true }
```
Comment (janhohenheim) [inline Cargo.toml]: Release mockiato and use the correct version

File: Cargo.toml [hunk 3]
```diff
@@ -35,7 +34,7 @@ default-features = false
 features = ["use-wasm-bindgen"]
 
 [dev-dependencies]
-mockiato = "0.4"
+mockiato = { git = "https://github.com/myelin-ai/mockiato" }
```
Comment (janhohenheim) [inline Cargo.toml]: Release mockiato and use the correct version

File: src/object.rs [hunk 4]
```diff
@@ -67,9 +77,15 @@ pub struct ObjectDescription {
...
     /// Arbitrary data associated with this object
-    pub associated_data: Vec<u8>,
+    #[serde(bound(deserialize = "T: AssociatedObjectData"))]
+    pub associated_data: T,
 }
 
+/// Required bounds for [`ObjectDescription::associated_data`].
+/// This trait is automatically implemented for all types that satisfy the supertraits.
+pub trait AssociatedObjectData =
+    Clone + Debug + PartialEq + Serialize + for<'de> Deserialize<'de> + 'static;
```
Comment (janhohenheim) [inline src/object.rs:87]:
```suggestion
    Clone + Debug + PartialEq + Serialize + for<'a> Deserialize<'a> + 'static;
```

File: src/object/object_behavior.rs [hunk 5]
```diff
@@ -28,9 +34,10 @@ pub trait ObjectBehaviorAsAny {
     fn as_any(&self) -> &'_ dyn Any;
 }
 
-impl<T> ObjectBehaviorAsAny for T
+impl<O, T> ObjectBehaviorAsAny<T> for O
```
Comment (janhohenheim) [inline src/object/object_behavior.rs:37]: Give `T` and `O` nice names

**Review by janhohenheim — CHANGES_REQUESTED** [review]

File: Cargo.toml (hunk 3 above)
Comment (tautropfli) [reply Cargo.toml]: Already in a todo

File: Cargo.toml (hunk 2 above)
Comment (tautropfli) [reply Cargo.toml]: Already in a todo

File: src/object.rs (hunk 4 above)
Comment (tautropfli) [reply src/object.rs:87]: The lifetime on the `Deserialize` trait is called `'de` thats why I've called it that too.

File: src/lib.rs (hunk 1 above)
Comment (tautropfli) [reply src/lib.rs]: Addressed in b4919d9

File: src/object/object_behavior.rs (hunk 5 above)
Comment (tautropfli) [reply src/object/object_behavior.rs:37]: Addressed in 4c08d22

**Review by janhohenheim — APPROVED** [review]

File: src/object/object_behavior.rs (hunk 5 above)
Comment (jeremystucki) [reply src/object/object_behavior.rs:37]: I think you missed some occurrences

**Pull request merged**